Training/Program.cs

```diff
@@ -1,33 +1,120 @@
+using static System.Console;
+using static Training.EState;
+namespace Training;
+
+class Program {
+   static void Main (string[] args) {
+      Dictionary<string, (string, string, string, string, string)> validPaths = new () {
+         { "C:\\WorkGIT\\Dakshita\\Training.sln", ("c", "workgit", "dakshita", "training.sln", ".sln")},
+         { "C:/words/words.txt", ("c", "words", "", "words.txt", ".txt")}
+      };
+      string[] invalidPaths = { "suppppp", "Ca:/workgit/hjh.txt", "c:\\Dakshita123.md", "C:\\workgit\\dakshita", "c:/dakshita.txt" };
+      foreach (var kvp in validPaths)
+         Validate (kvp.Key);
+      foreach (string path in invalidPaths)
+         Validate (path);
 
-      #region Operators ---------------------------------------------
-      #endregion
+      static void Validate (string path) =>
+         WriteLine (Path.PathParse (path, out (string, string, string, string, string) ActOutput) + "\t" + ActOutput);
+
+      //Path fp = new ("C:\\WorkGIT\\Dakshita\\Training.sln");
+      //WriteLine (fp);
+      //WriteLine (fp.FilePath);
+   }
+}
 
-      #region Implementation ----------------------------------------
-      #endregion
+#region class Path --------------------------------------------------------------------------
+public class Path {
+   #region Constructors ------------------------------------------
+   /// <summary>Creates new instance of a filepath</summary>
+   public Path (string path) => mPath = path.Trim ().ToLower ();
+   #endregion
 
-      #region Nested Types ------------------------------------------
-      #endregion
+   #region Properties ------------------------------------------
+   /// <summary>If path is valid, returns components of a file's path, else returns empty strings.</summary>
+   public (string Drive, string Dir, string Path, string Filename, string Ext) FilePath {
+      get {
+         // FilePath is computed only once for 1 instance of class (lazy evaluation).
+         if (mComputed)
+            return mFilePath;
+         else {
+            mValid = PathParse (mPath, out (string drive, string dir, string path, string filename, string ext) f);
+            mFilePath = mValid ? f : ("", "", "", "", "");
+            mComputed = true;
+            nComputations++;
+            return mFilePath;
+         }
+      }
+   }
+   string mPath;
+   (string drive, string dir, string path, string filename, string ext) mFilePath;
+   bool mComputed = false;
 
-      #region Private Data ------------------------------------------
-      #endregion 
+   /// <summary>Returns if current instance of Path is valid.</summary>
+   public bool IsValid {
+      get {
+         if (mComputed) return mValid;
+         else {
+            mValid = PathParse (mPath, out (string drive, string dir, string path, string filename, string ext) f);
+            mFilePath = mValid ? f : ("", "", "", "", "");
+            mComputed = true;
+            nComputations++;
+            return mValid;

```

Comment:
Below code is how we normally implement lazy evaluation where the code used to compute multiple vaiable is shared. In this case, Computing FilePath also compute valid state.
```suggestion
        _ = FilePath;
        return mValid;

```